### PR TITLE
process: describe Xen Troops development process

### PR DESCRIPTION
**Pull request by lorc**
This document is aimed to be our set of rules, which we will strictly follow. It should be considered as the final truth in any argument.

It describes important topics like "definition of done", licensing, submitting changes and so on.

### CODE REVIEW

File: process/process.md [hunk 1]
```diff
@@ -0,0 +1,295 @@
+# Preface
+
+This document defines processes adopted by Xen-Troops project. It
+should be followed by all Xen-Troops members. Parts of the document
+that describe commits, pull requests and such should be followed by
+external contributors as well.
+
+# Definition of Done
+
+Every task can is considered completed when resulting changes are
```
Comment (rshym) [inline process/process.md]: > can is 
Seems like misprint.

File: process/process.md [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+Also you can check our [contributing
+guide](https://xen-troops.github.io/contribution)
+
+Here we will re-iterate most important points and describe how it
+applies to GitHub, as we don't use mailing-list based review.
+
+## Coding style
+
+Every project has own coding style so we can't provide here any
+specific requirements. But you always should to adhere to the coding
```
Comment (rshym) [inline process/process.md]: s/"always should to adhere"/"should always adhere"

File: process/process.md [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+### Commit message format
+
+Commit message consist of one-line description (aka header), main
+description and tag lines. They are divided by empty lines:
+
+```
+subsystem: add possibility to grok baz
+
+When workin on foo, we noticed that blah-blah-blah.
```
Comment (rshym) [inline process/process.md]: s/workin/working

File: process/process.md [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+### Describe your Pull Request
+
+If your PR consists of one commit, GitHub will do this for you. But if
+there are more commits, you need to provide comprehensive description
+for the PR as a whole. The same rules apply: short and descriptive PR
+name, detailed description of a problem solved. Do not try to squash
+your commit messages, write description for the PR as one entity instead.
+
+If appropriate, write how you tested it. Wait automated checks to run.
```
Comment (rshym) [inline process/process.md]: Seems like s/"Wait automated"/"Wait for automated"

File: process/process.md [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+cases GitHub will assign reviewers immediately, but it is a good idea
+to ping reviewers to tell them that PR is really ready for review.
+
+Not every comment from reviewer require changes in code. But many of
+them are. If reviewer asks you a question, do not haste to change your
+code. Answer the question first.
+
+Every project has maintainer(s). Maintainers have full vision of
+project and its architecture. In most cases maintainers will support
+code you contributed for rest of it life. So maintainers can ask you
```
Comment (rshym) [inline process/process.md]: s/it/its

File: process/process.md [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+to make changes into your PR so it better fit into project's
+architecture. Even if your code is 100% and bug-free, it does not mean
```
Comment (rshym) [inline process/process.md]: > is 100% and bug-free
Seems something is missed here after "100%"

File: process/process.md [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+
+# Licensing
+
+All new code should be licensed under Apache-2.0 license by default.
+
+# "Temporary" changes
+
+In some cases it is tempting to make temporary fix to get project
+going forward. But practice shows that this temporary fixes tend to
+stick forever. So we trying to avoid them at all costs. But sometimes
```
Comment (rshym) [inline process/process.md]: s/"we trying"/"we are trying"

File: process/process.md [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+   will be done.
+
+4. Open issue for this change in the corresponding repository. Issue
+   should look like "Rework temporary fix introduced by PR
+   #NNN". Assign yourself to it. Set resolution date. Make sure that
+   PR page has reference to this issue.
+
+
+
+
```
Comment (rshym) [inline process/process.md]: Not needed empty lines.

**Comment by lorc** [comment]
Thanks @rshym, I have addressed your comments

**Comment by rshym** [comment]
`Reviewed-by: Ruslan Shymkevych <ruslan_shymkevych@epam.com>`

**Review by rshym — APPROVED** [review]

File: process/process.md [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+Every task is considered completed when resulting changes are fully
+integrated into the codebase. This means that you have done your task
+only **after** your Pull Request(s) was/were merged. Before that
+moment your task is **not** done.
+
+Not all tasks end as pull requests. For example, there can be research
+tasks where you research some topic or experiment with different
+approaches. Those still require some tangible results, like Wiki page
+for example.
+
```
Comment (klogg) [inline process/process.md:24]: Probably worth to mention that it is up to committers / maintainers to ensure code quality, test coverage, documentation completeness prior to merging the PR(s)

File: process/process.md [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+If you meet all of those requirements, you can ask integrator or
+maintainer to merge your PR.
+
+Please note, that we do not use merge commits. We follow rebase
+strategy.
+
+# Licensing
+
+All new code should be licensed under Apache-2.0 license by default.
```
Comment (klogg) [inline process/process.md]: "New code" above means completely new projects (new app, new lib, new daemon, etc.)
New code to existing projects (i.e. new modules for Xen, Linux kernel, etc.) shall follow the license of existing projects.
Use SPDX tags where possible: spdx.github.io/spdx-spec/v3.0/annexes/using-SPDX-short-identifiers-in-source-files/
For new files and significantly changed (new major features, major refactoring, etc.) add EPAM's copyright line in a suitable format.
If usure on licensing/copyright, ask Artem.

File: process/process.md [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+
+# "Temporary" changes
```
Comment (klogg) [inline process/process.md:354]: Better to move this section above "Check list" and have the "Issue created for fixing temporary change" item added to the check list

File: process/process.md [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+architecture. Even if your code is 100% correct and bug-free, it does
+not mean that it will be accepted as is.
+
+When all comments are addressed, reviewers need to explicitly approve
+pull request by providing Reviewed-by or Acked-by tag.
+
+After collecting all needed R-b tags, you should include them
+into appropriate commit messages, rebase changes onto current upstream
+branch and force push changes into your local branch.
+
```
Comment (klogg) [inline process/process.md:322]: I miss here rules to update pull request when addressing reviewer's comments. Somehow our Renesas colleagues had issue with this in the past: https://github.com/aoscloud/meta-aos-rcar-gen5/pull/2#issuecomment-1959032876

File: process/process.md [hunk 13]
```diff
@@ -0,0 +1,291 @@
+# Preface
+
+This document defines processes adopted by Xen-Troops project. It
```
Comment (klogg) [inline process/process.md]: Only "introducing code changes" processes. Other processes we have not yet developed (i.e. task breakdown/estimation, triage, ...)

File: process/process.md [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+Here we will re-iterate most important points and describe how it
+applies to GitHub, as we don't use mailing-list based review.
+
+## Coding style
+
+Every project has own coding style so we can't provide here any
+specific requirements. But you should always adhere to the coding
+style of a project where you making commit. When in doubt - look at
+surrounding code and follow its style.
+
```
Comment (klogg) [inline process/process.md:47]: Please suggest style to stick to when starting **new project(s)** in C, C++, Rust, Python, Bash. I really hate seeing CaMeLcAsE one day.

File: process/process.md [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+If project has a separate documentation files - check if your changes
+affect documented parts. Alter the documentation so it always
+up-to-date with the code. Do this in the same commit that changes the
+code. This is crucial.
+
+If you are adding public interfaces - document them in the same way,
+as other parts are documented or better. Always make sure that
+comment reflect the code.
+
```
Comment (klogg) [inline process/process.md:95]: Maybe start using Doxygen tags properly?

File: process/process.md [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+that describe commits, pull requests and such should be followed by
+external contributors as well.
+
+# Definition of Done
+
+Every task is considered completed when resulting changes are fully
+integrated into the codebase. This means that you have done your task
+only **after** your Pull Request(s) was/were merged. Before that
+moment your task is **not** done.
+
```
Comment (klogg) [inline process/process.md:14]: Sometimes DoD for us means code is in upstream. I suggest having a section on how to upstream (including mess with using private emails but submitting on behalf work emails)

**Comment by lorc** [comment]
Thanks, @klogg 

I tried to address your comments. Could you please take a view?

For ease of review I added a fix-up patch with new changes.

Also, I added this paragraph, based on today's discussion:

> Try to avoid unsolicited changes during review phase. Do not add new
features or change existing code unless asked by reviewer. This will
increase review time drastically, as reviewers will need to review the
PR after any such change.

**Comment by klogg** [comment]
`Reviewed-by: Artem Mygaiev <artem_mygaiev@epam.com>`

**Review by klogg — APPROVED** [review]

**Pull request merged**